src/nexmark/queries/q7.rs

```diff
@@ -0,0 +1,184 @@
+use super::NexmarkStream;
+use crate::{
+    nexmark::model::Event,
+    operator::{FilterMap, Min},
+    Circuit, OrdIndexedZSet, OrdZSet, Stream,
+};
+
+///
+/// Query 7: Highest Bid
+///
+/// What are the highest bids per period?
+///
+/// The original Nexmark Query7 calculate the highest bids in the last minute.
+/// We will use a shorter window (10 seconds) to help make testing easier.
+///
+/// ```sql
+/// CREATE TABLE discard_sink (
+///   auction  BIGINT,
+///   bidder  BIGINT,
+///   price  BIGINT,
+///   dateTime  TIMESTAMP(3),
+///   extra  VARCHAR
+/// ) WITH (
+///   'connector' = 'blackhole'
+/// );
+///
+/// INSERT INTO discard_sink
+/// SELECT B.auction, B.price, B.bidder, B.dateTime, B.extra
+/// from bid B
+/// JOIN (
+///   SELECT MAX(B1.price) AS maxprice, TUMBLE_ROWTIME(B1.dateTime, INTERVAL '10' SECOND) as dateTime
+///   FROM bid B1
+///   GROUP BY TUMBLE(B1.dateTime, INTERVAL '10' SECOND)
+/// ) B1
+/// ON B.price = B1.maxprice
+/// WHERE B.dateTime BETWEEN B1.dateTime  - INTERVAL '10' SECOND AND B1.dateTime;
+/// ```
+
+type Q7Stream = Stream<Circuit<()>, OrdZSet<(u64, u64, usize, u64, String), isize>>;
+type Q7Output = (u64, u64, usize, u64, String);
+
+const TUMBLE_SECONDS: u64 = 10;
+
+pub fn q7(input: NexmarkStream) -> Q7Stream {
+    // All bids indexed by date time to be able to window the result.
+    let bids_by_time: Stream<_, OrdIndexedZSet<u64, Q7Output, _>> =
+        input.flat_map_index(|event| match event {
+            Event::Bid(b) => Some((
+                b.date_time,
+                (b.auction, b.bidder, b.price, b.date_time, b.extra.clone()),
+            )),
+            _ => None,
+        });
+
+    // Similar to the sliding window of q5, we want to find the largest timestamp
+    // from the input stream for the current time, with the window ending at the
+    // previous 10 second multiple.
+    // Set the watermark to `TUMBLE_SECONDS` in the past.
+    let watermark = bids_by_time.watermark_monotonic(|date_time| date_time - TUMBLE_SECONDS * 1000);
```

Comment:
We should use `WATERMARK` constant here, which should probably match the out-of-ordedness bounds used by the generator.